File: moviejava/app/src/main/java/com/example/moviejava/model/State.java
package com.example.moviejava.model;

public class State<T> {
    private final Status status;
    private final T data;
    private final Throwable throwable;

    public State(Status status, T data, Throwable throwable) {
        this.status = status;
        this.data = data;
        this.throwable = throwable;
    }

    public Status getStatus() {
        return status;
    }

    public T getData() {
        return data;
    }

    public Throwable getThrowable() {
        return throwable;
    }

    public boolean isLoading() {
        return status == Status.Loading;
    }

    public boolean isSuccess() {
        return status == Status.Success;
    }

    public boolean isError() {
        return status == Status.Error;
    }

    public boolean isFinish() {
        return isError() || isSuccess();
    }

    public static <T> State<T> success(T data) {
        return new State<>(Status.Success, data, null);
    }

    public static <T> State<T> error(Throwable throwable) {
        return new State<>(Status.Error, null, throwable);
    }

    public static <T> State<T> loading() {
        return new State<>(Status.Loading, null, null);
    }
}

enum Status {
    Loading, Success, Error
}

File: movie/app/src/main/java/com/example/movie/model/State.kt
package com.example.movie.model


class State<T>(
    val status: Status, val data: T? = null, val throwable: Throwable? = null
) {


    val isLoading = status is Status.Loading
    val isSuccess = status is Status.Success
    val isError = status is Status.Error
    val isFinish: Boolean
        get() = isError || isSuccess

    companion object {
        fun <T> success(data: T? = null): State<T> {
            return State(Status.Success, data)
        }

        fun <T> error(throwable: Throwable?): State<T> {
            return State(Status.Error, throwable = throwable)
        }

        fun <T> loading(): State<T> {
            return State(Status.Loading)
        }
    }

    sealed class Status {
        object Loading : Status()
        object Success : Status()
        object Error : Status()
    }
}
